### ISSUE TITLE
`parenthesized_params_in_types_and_modules` future-compatibility warnings

### ISSUE THREAD

**Issue by qnighy** [rust-lang/rust#42238]
This is the **summary issue** for the `PARENTHESIZED_PARAMS_IN_TYPES_AND_MODULES` future-compatibility warning and other related errors. The goal of this page is describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made.

#### What is the warning for?

The `parenthesized_params_in_types_and_modules` warning is issued when you use parenthesized generic parameters in modules, primitive types, type parameters, or `Self`. For structs, enums, enum variants, unions, and type aliases, the syntax has already issued an error. Typical example of this warning is a code like this:

```rust
fn something(&self) -> i32 () {
  //                       ^^ parenthesized params

  ..
  let x = self.foo() as usize();
  //                         ^^ parenthesized params
  ..
}
```

This warning can always be fixed by removing the parentheses.

#### When will this warning become a hard error?
                    
At the beginning of each 6-week release cycle, the Rust compiler team will review the set of outstanding future compatibility warnings and nominate some of them for **Final Comment Period**. Toward the end of
the cycle, we will review any comments and make a final determination whether to convert the warning into a hard error or remove it entirely.

#### Current status
- [x] https://github.com/rust-lang/rust/pull/41856 introduces the `parenthesized_params_in_types_and_modules` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/42894 makes the `parenthesized_params_in_types_and_modules` lint deny-by-default
- [ ] PR ? makes the `parenthesized_params_in_types_and_modules` lint a hard error

**Comment by qnighy** [comment]
I noticed there is a case where `()` cannot just be removed.

```rust
fn main() {
    0 as usize() < 1; // `()` can't just be removed.
}
```

The cause is the well-known ambiguity between "less-than" and "open angle bracket". In this case, you can fix the problem like this:

```rust
fn main() {
    0 as (usize) < 1; // OK
    (0 as usize) < 1; // OK
}
```

**Comment by tchamelot** [comment]
I don't know if this should go here, I am new to Rust and  I still make mistakes such as confuse ":" and "=" for struct. Today, the compiler brought me here through a link because of this error. Is this the standard behavior for this kind of error.

`error: parenthesized parameters may only be used with a trait                                  
   --> src/main.rs:100:28                                                                      
    |                                                                                          
100 |     let mut ally : Pod::new();                                                           
    |                            ^^                                                            
    |                                                                                          
    = note: #[deny(parenthesized_params_in_types_and_modules)] on by default                   
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #42238 <https://github.com/rust-lang/rust/issues/42238>
`

**Comment by earthengine** [comment]
This might be needed, but the current implementation have problems as `()` is the ony way to express the tuple types right now, but a type like `Result<(),()>` were still reported `deny`.

Check this playground link.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ea8942b0cd702ae1ed586a67963f0bf9


Please, give a workaround or relax this rule to allow empty tuple type. This is a breaking change.
